### DISCUSSION TITLE
how to use (install) fastapi in a yocto based image

### DISCUSSION THREAD

**Discussion by svhb1000** [fastapi/fastapi#12255] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
Bitbake Recipe to (try) install fastapi : 

SUMMARY = "Recipe to embedded the Python PiP Package fastapi"
HOMEPAGE ="https://pypi.org/project/fastapi"
LICENSE = "MIT"
LIC_FILES_CHKSUM = "file://LICENSE;md5=95792ff3fe8e11aa49ceb247e66e4810"

inherit pypi python_setuptools_build_meta

DEPENDS += " python3-poetry-core python3-pdm python3-pdm-backend"

UPSTREAM_CHECK_URI = "https://pypi.python.org/pypi/fastapi"

SRC_URI[sha256sum] = "f93b4ca3529a8ebc6fc3fcf710e5efa8de3df9b41570958abf1d97d843138004"
```

### Description

This is maybe a bit a strange question here, but I find no usefull information elsewhere.
For an embedded system I'm setting up, I use Yocto and OpenEmbedded to create an embedded linux distro.

Anyone familiar with this? How to set up a recipe to build without the error "ERROR Backend 'pdm.backend' is not available".

Output from above recipe description : 

ERROR: fastapi-0.115.0-r0 do_compile: ExecutionError('/mnt/develop/yocto/poky/build/tmp/work/core2-64-poky-linux/fastapi/0.115.0/temp/run.do_compile.3336017', 1, None, None)
ERROR: Logfile of failure stored in: /mnt/develop/yocto/poky/build/tmp/work/core2-64-poky-linux/fastapi/0.115.0/temp/log.do_compile.3336017
Log data follows:
| DEBUG: Executing shell function do_compile
| * Getting build dependencies for wheel...
| 
| Traceback (most recent call last):
|   File "/mnt/develop/yocto/poky/build/tmp/work/core2-64-poky-linux/fastapi/0.115.0/recipe-sysroot-native/usr/lib/python3.12/site-packages/pyproject_hooks/_impl.py", line 321, in _call_hook
|     raise BackendUnavailable(data.get('traceback', ''))
| pyproject_hooks._impl.BackendUnavailable: Traceback (most recent call last):
|   File "/mnt/develop/yocto/poky/build/tmp/work/core2-64-poky-linux/fastapi/0.115.0/recipe-sysroot-native/usr/lib/python3.12/site-packages/pyproject_hooks/_in_process/_in_process.py", line 77, in _build_backend
|     obj = import_module(mod_path)
|           ^^^^^^^^^^^^^^^^^^^^^^^
|   File "/mnt/develop/yocto/poky/build/tmp/work/core2-64-poky-linux/fastapi/0.115.0/recipe-sysroot-native/usr/lib/python3.12/importlib/__init__.py", line 90, in import_module
|     return _bootstrap._gcd_import(name[level:], package, level)
|            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
|   File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
|   File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
|   File "<frozen importlib._bootstrap>", line 1310, in _find_and_load_unlocked
|   File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
|   File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
|   File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
|   File "<frozen importlib._bootstrap>", line 1324, in _find_and_load_unlocked
| ModuleNotFoundError: No module named 'pdm'
| 
**| ERROR Backend 'pdm.backend' is not available.**
| WARNING: exit code 1 from a shell command.
ERROR: Task (/mnt/develop/yocto/poky/meta-mylayer/recipes-fastapi/fastapi/fastapi_0.115.0.bb:do_compile) failed with exit code '1'
NOTE: Tasks Summary: Attempted 3766 tasks of which 3745 didn't need to be rerun and 1 failed.

### Operating System

Other

### Operating System Details

Yocto build (core-image-minimal)

### FastAPI Version

0.115

### Pydantic Version

na

### Python Version

3.12

**Comment by ppilz** [comment]
In case this is still relevant to you (or anyone else for that matter): After I started doing the work myself, I have just noticed that there is a patch currently in review to include FastAPI in Yocto in meta-python: [Yocto patchwork - Add FastAPI](https://patchwork.yoctoproject.org/project/oe/patch/20250423143607.453363-1-fleuryhugo7@gmail.com/)
The error `ERROR Backend 'pdm.backend is not available` stems from not using `python3-pdm-native` and `python3-pdm-backend-native` - note the -native suffix.

Side-note: I am on the Scarthgap branch of Yocto currently and there are a couple of recipes for dependencies missing since they are only in master, so one would have to include those as well:

- python3-starlette
- python3-outcome
- python3-multipart
- python3-trio

On master branch you should be fine though.

**Comment by YuriiMotov** [comment]
Let's close this for now to keep the discussions clean.
If you’ve found a solution, please share it here.
If you still need help, feel free to let us know.
